### PR TITLE
Add nvbench environment class for initializing RMM in benchmarks

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Adds an environment class to initialize the RMM pool memory resource manager before running benchmarks through nvbench. This removes the need to initialize RMM on every benchmark call which improves benchmark run performance and fixes some GPU metrics gathering when run under nsys.

This requires a patch to the nvbench source to enable this feature and is part of the following pull request:  https://github.com/NVIDIA/nvbench/pull/123
The patch can be removed once the PR is merged and source available to the libcudf build.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
@davidwendt did you want to merge this PR and revert the patch once the nvbench PR is merged, or wait for the nvbench PR? I'm fine either way; I know that the dev cycle on nvbench PRs is slower than ours.

File: cpp/benchmarks/fixture/rmm_pool_raii.hpp
Comment (vyasr) [inline cpp/benchmarks/fixture/rmm_pool_raii.hpp:83]: Could you use the `rmm_pool_raii` directly here? If the goal is to restrict the scope to more nvbench-specific things, the alternative question is, does `rmm_pool_raii` have any other purpose or could we just rename it to `nvbench_base_fixture`?

Comment (vyasr) [reply cpp/benchmarks/fixture/rmm_pool_raii.hpp:83]: As an aside, while we're making this change should we change the namespace? Our GTest fixture is in cudf::test, maybe this object should live in cudf::benchmark.

Comment (davidwendt) [reply cpp/benchmarks/fixture/rmm_pool_raii.hpp:83]: I wanted to just wrap in case we wanted add some other specific logic. I'd assume it had some other purpose. Regardless I felt it was clearer in its desired intent in the current implementation.
There is no `cudf::benchmark` namespace. I think creating the namespace would be out of scope for this PR.

**Comment by davidwendt** [comment]
> @davidwendt did you want to merge this PR and revert the patch once the nvbench PR is merged, or wait for the nvbench PR? I'm fine either way; I know that the dev cycle on nvbench PRs is slower than ours.

I did not want to wait for the nvbench PR to be merged first.

**Review by ttnghia — APPROVED** [review]
So does this mean this PR depends on https://github.com/NVIDIA/nvbench/pull/123? Look like that #123 is ready to merge.

**Review by bdice — APPROVED** [review]
Approved. Let’s try to note a “fixed in” version if we know what we might expect.

File: cpp/cmake/thirdparty/patches/thrust_override.json
Comment (davidwendt) [reply cpp/cmake/thirdparty/patches/thrust_override.json:39]: Not until it is merged and that does not look to be anytime soon.

File: cpp/cmake/thirdparty/patches/thrust_override.json
Comment (robertmaynard) [inline cpp/cmake/thirdparty/patches/thrust_override.json:36]: This shouldn't be in the Thrust patches override file as it will be applied when Thrust is downloaded and not when nvbench is.

We will also want to make a second file ( nvbench_override ) and have just the new entry in it:

```
{
  "packages" : {
    "nvbench" : {
      "patches" : [        
        {
          "file" : "${current_json_dir}/nvbench_global_setup.diff",
          "issue" : "Fix add support for global setup to initialize RMM in nvbench [https://github.com/NVIDIA/nvbench/pull/123]",
          "fixed_in" : ""
        }
      ]
    }
  }
}
```

We will want to add a `get_nvbench.cmake` and have it do the following logic before call `rapids_cpm_nvbench`:
```cmake
  include(${rapids-cmake-dir}/cpm/nvbench.cmake)
  include(${rapids-cmake-dir}/cpm/package_override.cmake)

  set(cudf_patch_dir "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/patches")
  rapids_cpm_package_override("${cudf_patch_dir}/nvbench_override.json")
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: cpp/benchmarks/fixture/rmm_pool_raii.hpp
Comment (karthikeyann) [inline cpp/benchmarks/fixture/rmm_pool_raii.hpp:79]: Is it possible to inherit this from existing `NVBENCH_ENVIRONMENT` ?
This would allow global fixture to be layered.

File: cpp/CMakeLists.txt
Comment (robertmaynard) [inline cpp/CMakeLists.txt:788]: Missed this the first time, can you move the following over to the function inside `get_nvbench`:
```cmake
  # do not support the latest version of fmt, which is automatically pulled into our conda
  # environments by mamba.
  set(CPM_DOWNLOAD_fmt TRUE)
```

File: cpp/cmake/thirdparty/patches/nvbench_override.json
Comment (robertmaynard) [inline cpp/cmake/thirdparty/patches/nvbench_override.json:4]:
```suggestion
    "nvbench" : {
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]
CMake changes lgtm

File: cpp/benchmarks/fixture/rmm_pool_raii.hpp
Comment (davidwendt) [reply cpp/benchmarks/fixture/rmm_pool_raii.hpp:79]: This would be a problem since the `NVBENCH_ENVIRONMENT` needs to be set before including `nvbench.cuh` so we'd get a circular dependency. I'd rather nvbench support layering first and then use that.
Just curious but what is the use case for layering?

Comment (karthikeyann) [reply cpp/benchmarks/fixture/rmm_pool_raii.hpp:79]:
```suggestion
#ifndef NVBENCH_ENVIRONMENT
struct nvbench_base_fixture {
  rmm_pool_raii _mr;
};
#else
struct nvbench_base_fixture : NVBENCH_ENVIRONMENT {
  rmm_pool_raii _mr;
};
#undef NVBENCH_ENVIRONMENT
#endif
```

Comment (davidwendt) [reply cpp/benchmarks/fixture/rmm_pool_raii.hpp:79]: I do not agree with this change for the reasons given previously.

**Review by robertmaynard — APPROVED** [review]
CMake changes are now great :+1:

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

**Pull request merged**
